### DISCUSSION TITLE
How to publish to Amazon Appstore using old RSA-1024 certificate?

### DISCUSSION THREAD

**Discussion by OMA2k** [airsdk/Adobe-Runtime-Support#2668] [Q&A] [answered]
Recently I came across an issue when trying to update a decade-old app in the Play Store. The .p12 certificate it used was RSA-1024, which wasn't supported by Java 11, which is used by ADT in the latest versions of the AIR SDK. Thanks to your help I could solve this by creating and providing an upload certificate to create AAB packages for the Play Store. I described this [here](https://github.com/airsdk/Adobe-Runtime-Support/discussions/2659#discussioncomment-5972946).

Now I want to upload an update for the app in the Amazon Appstore as well. This store also allows AAB packages to be used, but it doesn't have an option to use an "upload certificate" so I can use an RSA-2048, like the Play Store, or at least I haven't found it.

Is my only option to use the old RSA-1024 certificate with Java 8u112 to generate APKs for the Amazon Appstore instead of using Java 11 or higher to generate AAB packages?

**Comment by ajwfrost** [comment]
Would it be feasible to use a new 2048-bit certificate to create an initial APK or AAB file, but then re-sign it using Java 8u112 and the "apksigner" or "jarsigner" applications respectively, with your 1024-bit certificate?

I think it's just the final package which is signed, so it should be possible to do this as a second step I hope!

**Comment by marchbold** [comment]
Doesn't Amazon already resign all applications with an Amazon provided certificate and the certificate you are using is just an "upload certificate"? 

So I believe you can just change it without any issues here.

**Reply by OMA2k** [reply]
So it doesn't matter whatever certificate I use to compile because Amazon will re-sign the package anyway automatically?
I didn't know that and didn't find that in the Amazon documentation I read.

Well, before reading this last comment, I followed @ajwfrost's suggestion and tried to use "apksigner", but it returned some errors because apparently it can't properly detect some parameters of AAB files, so I went with "jarsigner" instead, and did this:

```
export JAVA_HOME=`/usr/libexec/java_home -v1.8.0_112`
export JAVACMD=${JAVA_HOME}/bin/java
jarsigner -storetype pkcs12 -keystore old_cert.p12 -tsa "http://timestamp.digicert.com" -storepass xxxx app_package.aab 1
```

I had to add a timestamp server URL for it to work with no warnings. The 1 at the end is the alias for ADT created self-signed certificates.

The Amazon Appstore accepted the AAB file, but I guess this AAB re-signing step is unnecessary because Amazon apparently re-signs the packages again automatically and accepts any certificate for package upload. Good to know.

**Reply by marchbold** [reply, marked as answer]
https://developer.amazon.com/docs/app-submission/understanding-submission.html#app_signatures

> When you submit your app, Amazon removes the signature you used to sign your app and re-signs it with an Amazon signature that is unique to you, does not change, and is the same for all apps in your account.

This is basically what Google are now doing with the play app signing stuff, so I think this is one (of the few) areas that amazon where ahead of Google Play. 

A recent release we did we changed the certificate without issues, so I'd say you are going to a lot of effort for no need :)

**Reply by OMA2k** [reply]
Thanks for pointing to the relevant part of the docs. So this is, definitely, easier than I thought.
